arlo: Add alarm control panel component

vickyg3 opened the pull request:
## Description:

Allows importing arlo base stations as an alarm control panel
component in HA. Lets the users configure a custom home mode since
arlo does not have a built-in home mode.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):**

https://github.com/home-assistant/home-assistant.github.io/pull/3535

## Example entry for `configuration.yaml` (if applicable):
```yaml
alarm_control_panel:
  - platform: arlo
    home_mode_name: Home
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

vickyg3 committed:
arlo: Add alarm control panel component

Allows importing arlo base stations as an alarm control panel
component in HA. Lets the users configure a custom home mode since
arlo does not have a built-in home mode.

vickyg3 committed:
fix lint and houndci comments

MartinHjelmare commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,109 @@
+DISARMED = "disarmed"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOME_MODE_NAME, default=ARMED): cv.string,
```
Is this configurable on the Arlo device? Is that why this should be configurable here?

vickyg3 replied:
Arlo base station only has two modes by default. Armed and Disarmed. But users can create custom modes on the arlo app. Since HA alarm panel component has the notion of 'home' mode, this configuration just let's them map one of their 'custom' modes as 'home' mode. If they don't have one, then home and away modes are the same (and the status will always say armed_away).

Sorry if i wasn't clear in the description. :)

vickyg3 committed:
Use async_update to update the state

Move the state updating code from state() to update() since it does
I/O.

vickyg3 commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,109 @@
+        num_retries = 3
+        i = 0
+        while i < num_retries:
+            mode = self._base_station.mode
```
Thank you for pointing this out. Yes it does I/O. Have moved it to the async_update method.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,117 @@
+        """Initialize the alarm control panel."""
+        self._base_station = data
+        self._home_mode_name = home_mode_name
+        self._state = self._base_station.mode
```
Init this as `None`, and make sure update is called after adding entities, like this:
```py
async_add_devices(devices, True)
```

vickyg3 replied:
Done.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,117 @@
+            mode = self._base_station.mode
+            if mode:
+                self._state = self._get_state_from_mode(mode)
+                return self._state
```
Don't return anything, just return.

vickyg3 replied:
Done.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,117 @@
+                return self._state
+            i += 1
+        self._state = None
+        return self._state
```
Remove.

vickyg3 replied:
Done.

vickyg3 committed:
Do not set state in __init__

Make sure that update is called by passing the second parameter to
async_add_devices.

vickyg3 reviewed (commented):
Have made all the changes you requested. Please take another look. Thanks!

MartinHjelmare commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,116 @@
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
```
Why this?

MartinHjelmare reviewed (approved):
Looks good!

vickyg3 commented:
Thanks for the review!

fabaff commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,116 @@
+                                                          PLATFORM_SCHEMA)
+from homeassistant.const import (ATTR_ATTRIBUTION, STATE_ALARM_ARMED_AWAY,
+                                 STATE_ALARM_ARMED_HOME, STATE_ALARM_DISARMED)
+import homeassistant.helpers.config_validation as cv
```
Please keep the import [ordered](https://home-assistant.io/developers/development_guidelines/).

fabaff commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,116 @@
+"""
+This component provides HA alarm_control_panel support for Arlo System.
```
`Support for Arlo...` is OK.

fabaff commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,116 @@
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up a sensor for an Arlo device."""
```
Stale docstring.

fabaff commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,116 @@
+
+
+class ArloBaseStation(AlarmControlPanel):
+    """An alarm_control_panel implementation for Arlo."""
```
Please remove the underscores.

fabaff commented on homeassistant/components/alarm_control_panel/arlo.py:
```diff
@@ -0,0 +1,116 @@
+            'device_id': self._base_station.device_id
+        }
+
+    def _get_state_from_mode(self, mode):
```
Docstring is missing.

vickyg3 committed:
Order imports and fix dos-strings

vickyg3 reviewed (commented):
i've fixed things that you suggested. PTAL. Thanks!

fabaff commented:
Most of my comments weren't addressed. But instead of adding the same details again I will merged it and fix then.

vickyg3 commented:
I'm sorry. I thought i addressed all your comments. Can you please point out which ones weren't addressed? I'm happy to send another PR to fix those.
